`panic` on geth/downloader/beaconsync.go:220

version `Geth/v1.11.0-unstable-722bb210-20221117/linux-amd64/go1.19.3`

Panic on the disco
```
 Dec 04 14:06:30 bootnode-azure-brazilsouth-001 geth WARN [12-04|13:06:29.937] Snapshot is not continuous with chain snaproot=9ac2ed..16d504 chainroot=e459a1..23facd
Dec 04 14:06:30 bootnode-azure-brazilsouth-001 geth INFO [12-04|13:06:29.937] Loaded local transaction journal transactions=0 dropped=0
Dec 04 14:06:30 bootnode-azure-brazilsouth-001 geth INFO [12-04|13:06:29.937] Regenerated local transaction journal transactions=0 accounts=0
Dec 04 14:06:30 bootnode-azure-brazilsouth-001 geth WARN [12-04|13:06:29.937] Switch sync mode from snap sync to full sync
Dec 04 14:06:30 bootnode-azure-brazilsouth-001 geth INFO [12-04|13:06:29.937] Chain post-merge, sync via beacon client
Dec 04 14:06:30 bootnode-azure-brazilsouth-001 geth INFO [12-04|13:06:29.938] Gasprice oracle is ignoring threshold set threshold=2
Dec 04 14:06:30 bootnode-azure-brazilsouth-001 geth WARN [12-04|13:06:29.938] Old unclean shutdowns found count=1593
Dec 04 14:06:30 bootnode-azure-brazilsouth-001 geth WARN [12-04|13:06:29.938] Unclean shutdown detected booted=2022-12-04T13:04:18+0000 age=2m11s
Dec 04 14:06:30 bootnode-azure-brazilsouth-001 geth WARN [12-04|13:06:29.938] Unclean shutdown detected booted=2022-12-04T13:04:30+0000 age=1m59s
Dec 04 14:06:30 bootnode-azure-brazilsouth-001 geth WARN [12-04|13:06:29.938] Unclean shutdown detected booted=2022-12-04T13:04:42+0000 age=1m47s
Dec 04 14:06:30 bootnode-azure-brazilsouth-001 geth WARN [12-04|13:06:29.938] Unclean shutdown detected booted=2022-12-04T13:04:53+0000 age=1m36s
Dec 04 14:06:30 bootnode-azure-brazilsouth-001 geth WARN [12-04|13:06:29.938] Unclean shutdown detected booted=2022-12-04T13:05:06+0000 age=1m23s
Dec 04 14:06:30 bootnode-azure-brazilsouth-001 geth WARN [12-04|13:06:29.938] Unclean shutdown detected booted=2022-12-04T13:05:17+0000 age=1m12s
Dec 04 14:06:30 bootnode-azure-brazilsouth-001 geth WARN [12-04|13:06:29.938] Unclean shutdown detected booted=2022-12-04T13:05:29+0000 age=1m
Dec 04 14:06:30 bootnode-azure-brazilsouth-001 geth WARN [12-04|13:06:29.938] Unclean shutdown detected booted=2022-12-04T13:05:42+0000 age=47s
Dec 04 14:06:30 bootnode-azure-brazilsouth-001 geth WARN [12-04|13:06:29.938] Unclean shutdown detected booted=2022-12-04T13:05:53+0000 age=36s
Dec 04 14:06:30 bootnode-azure-brazilsouth-001 geth WARN [12-04|13:06:29.938] Unclean shutdown detected booted=2022-12-04T13:06:05+0000 age=24s
Dec 04 14:06:30 bootnode-azure-brazilsouth-001 geth WARN [12-04|13:06:29.938] Unclean shutdown detected booted=2022-12-04T13:06:17+0000 age=12s
Dec 04 14:06:30 bootnode-azure-brazilsouth-001 geth INFO [12-04|13:06:29.938] Allocated cache and file handles database=/root/.ethereum/geth/les.server cache=16.00MiB handles=16
Dec 04 14:06:30 bootnode-azure-brazilsouth-001 geth INFO [12-04|13:06:29.947] Configured checkpoint oracle address=0x9a9070028361F7AAbeB3f2F2Dc07F82C4a98A02a signers=5 threshold=2
Dec 04 14:06:30 bootnode-azure-brazilsouth-001 geth INFO [12-04|13:06:29.948] Loaded latest checkpoint section=490 head=96c7bd..ab961d chtroot=532f7c..e64252 bloomroot=07d2a8..68e346
Dec 04 14:06:30 bootnode-azure-brazilsouth-001 geth WARN [12-04|13:06:29.948] Engine API enabled protocol=eth
Dec 04 14:06:30 bootnode-azure-brazilsouth-001 geth INFO [12-04|13:06:29.948] Starting peer-to-peer node instance=Geth/v1.11.0-unstable-722bb210-20221117/linux-amd64/go1.19.3
Dec 04 14:06:30 bootnode-azure-brazilsouth-001 geth INFO [12-04|13:06:29.964] Mapped network port proto=tcp extport=30303 intport=30303 interface=ExtIP(191.234.162.198)
Dec 04 14:06:30 bootnode-azure-brazilsouth-001 geth INFO [12-04|13:06:29.964] Mapped network port proto=udp extport=30303 intport=30303 interface=ExtIP(191.234.162.198)
Dec 04 14:06:30 bootnode-azure-brazilsouth-001 geth INFO [12-04|13:06:30.062] New local node record seq=3041 id=32f7e89d30e83b68 ip=191.234.162.198 udp=30303 tcp=30303
Dec 04 14:06:30 bootnode-azure-brazilsouth-001 geth INFO [12-04|13:06:30.203] Started P2P networking self=enode://103858bdb88756c71f15e9b5e09b56dc1be52f0a5021d46301dbbfb7e130029cc9d0d6f73f693bc29b665770fff7da4d34f3c6379fe12721b5d7a0bcb5ca1fc1@191.234.162.198:30303
Dec 04 14:06:30 bootnode-azure-brazilsouth-001 geth INFO [12-04|13:06:30.205] IPC endpoint opened url=/root/.ethereum/geth.ipc
Dec 04 14:06:30 bootnode-azure-brazilsouth-001 geth INFO [12-04|13:06:30.205] Loaded JWT secret file path=/execution-auth.jwt crc32=0xba247cd0
Dec 04 14:06:30 bootnode-azure-brazilsouth-001 geth INFO [12-04|13:06:30.205] WebSocket enabled url=ws://[::]:8551
Dec 04 14:06:30 bootnode-azure-brazilsouth-001 geth INFO [12-04|13:06:30.205] HTTP server started endpoint=[::]:8551 auth=true prefix= cors=localhost vhosts=*
Dec 04 14:06:30 bootnode-azure-brazilsouth-001 geth INFO [12-04|13:06:30.206] Stats daemon started
Dec 04 14:06:37 bootnode-azure-brazilsouth-001 geth WARN [12-04|13:06:36.807] Ignoring payload with missing parent number=16,111,632 hash=e9c1fd..f8ab50 parent=fe62f5..69bae3
Dec 04 14:06:37 bootnode-azure-brazilsouth-001 geth INFO [12-04|13:06:36.924] Forkchoice requested sync to new head number=16,111,632 hash=e9c1fd..f8ab50
Dec 04 14:06:37 bootnode-azure-brazilsouth-001 geth INFO [12-04|13:06:36.926] Block synchronisation started
Dec 04 14:06:37 bootnode-azure-brazilsouth-001 geth panic: runtime error: invalid memory address or nil pointer dereference
Dec 04 14:06:37 bootnode-azure-brazilsouth-001 geth [signal SIGSEGV: segmentation violation code=0x1 addr=0x1c0 pc=0xad3ac8]
Dec 04 14:06:37 bootnode-azure-brazilsouth-001 geth goroutine 6345 [running]:
Dec 04 14:06:37 bootnode-azure-brazilsouth-001 geth github.com/ethereum/go-ethereum/eth/downloader.(*Downloader).findBeaconAncestor(0xc000122c00)
Dec 04 14:06:37 bootnode-azure-brazilsouth-001 geth github.com/ethereum/go-ethereum/eth/downloader/beaconsync.go:220 +0xe8
Dec 04 14:06:37 bootnode-azure-brazilsouth-001 geth github.com/ethereum/go-ethereum/eth/downloader.(*Downloader).syncWithPeer(0xc000122c00, 0x0, {0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...}, ...)
Dec 04 14:06:37 bootnode-azure-brazilsouth-001 geth github.com/ethereum/go-ethereum/eth/downloader/downloader.go:527 +0x911
Dec 04 14:06:37 bootnode-azure-brazilsouth-001 geth github.com/ethereum/go-ethereum/eth/downloader.(*Downloader).synchronise(0x3?, {0x0, 0x0}, {0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...}, ...)
Dec 04 14:06:37 bootnode-azure-brazilsouth-001 geth github.com/ethereum/go-ethereum/eth/downloader/downloader.go:439 +0x512
Dec 04 14:06:37 bootnode-azure-brazilsouth-001 geth github.com/ethereum/go-ethereum/eth/downloader.(*beaconBackfiller).resume.func1()
Dec 04 14:06:37 bootnode-azure-brazilsouth-001 geth github.com/ethereum/go-ethereum/eth/downloader/beaconsync.go:108 +0x96
Dec 04 14:06:37 bootnode-azure-brazilsouth-001 geth created by github.com/ethereum/go-ethereum/eth/downloader.(*beaconBackfiller).resume
Dec 04 14:06:37 bootnode-azure-brazilsouth-001 geth github.com/ethereum/go-ethereum/eth/downloader/beaconsync.go:98 +0x136
```

This is the very first of the panics, which did not happen in conjuction with a reboot:
```
WARN [12-04|12:03:59.795] Peer delivering stale transactions peer=e8012de3e7972685711cafbf6fad4e4080aa4f64536341194ee0bf0eb4372b01 rejected=50
INFO [12-04|12:04:00.172] Forkchoice requested sync to new head number=16,111,322 hash=d0c18d..8b964d
WARN [12-04|12:04:00.230] Peer delivering stale transactions peer=651ba08ecbfd7d16b6a43bae82ffd54c31fb4663145c469d7c7e4b0777077374 rejected=41
WARN [12-04|12:04:00.428] Peer delivering stale transactions peer=e8012de3e7972685711cafbf6fad4e4080aa4f64536341194ee0bf0eb4372b01 rejected=54
WARN [12-04|12:04:00.802] Peer delivering stale transactions peer=651ba08ecbfd7d16b6a43bae82ffd54c31fb4663145c469d7c7e4b0777077374 rejected=43
INFO [12-04|12:04:01.100] Imported new chain segment number=16,111,306 hash=95fa2a..39901e blocks=5 txs=704 mgas=70.796 elapsed=13.613s mgasps=5.200 age=3m14s dirty=993.03MiB
WARN [12-04|12:04:01.275] Peer delivering stale transactions peer=e8012de3e7972685711cafbf6fad4e4080aa4f64536341194ee0bf0eb4372b01 rejected=56
WARN [12-04|12:04:01.828] Peer delivering stale transactions peer=651ba08ecbfd7d16b6a43bae82ffd54c31fb4663145c469d7c7e4b0777077374 rejected=36
WARN [12-04|12:04:02.343] Peer delivering stale transactions peer=e8012de3e7972685711cafbf6fad4e4080aa4f64536341194ee0bf0eb4372b01 rejected=53
INFO [12-04|12:04:03.184] Imported new chain segment number=16,111,307 hash=af366d..bd916b blocks=1 txs=267 mgas=22.671 elapsed=2.084s mgasps=10.876 age=3m4s dirty=994.42MiB
INFO [12-04|12:04:03.185] Downloader queue stats receiptTasks=0 blockTasks=0 itemSize=79.14KiB throttle=3313
WARN [12-04|12:04:03.260] Peer delivering stale transactions peer=651ba08ecbfd7d16b6a43bae82ffd54c31fb4663145c469d7c7e4b0777077374 rejected=45
WARN [12-04|12:04:04.758] Peer delivering stale transactions peer=651ba08ecbfd7d16b6a43bae82ffd54c31fb4663145c469d7c7e4b0777077374 rejected=36
WARN [12-04|12:04:05.198] Peer delivering stale transactions peer=651ba08ecbfd7d16b6a43bae82ffd54c31fb4663145c469d7c7e4b0777077374 rejected=44
WARN [12-04|12:04:05.463] Peer delivering stale transactions peer=e8012de3e7972685711cafbf6fad4e4080aa4f64536341194ee0bf0eb4372b01 rejected=52
WARN [12-04|12:04:05.822] Peer delivering stale transactions peer=e8012de3e7972685711cafbf6fad4e4080aa4f64536341194ee0bf0eb4372b01 rejected=33
WARN [12-04|12:04:06.156] Peer delivering stale transactions peer=651ba08ecbfd7d16b6a43bae82ffd54c31fb4663145c469d7c7e4b0777077374 rejected=33
WARN [12-04|12:04:06.522] Peer delivering stale transactions peer=651ba08ecbfd7d16b6a43bae82ffd54c31fb4663145c469d7c7e4b0777077374 rejected=34
WARN [12-04|12:04:07.115] Peer delivering stale transactions peer=651ba08ecbfd7d16b6a43bae82ffd54c31fb4663145c469d7c7e4b0777077374 rejected=54
WARN [12-04|12:04:07.490] Peer delivering stale transactions peer=651ba08ecbfd7d16b6a43bae82ffd54c31fb4663145c469d7c7e4b0777077374 rejected=44
WARN [12-04|12:04:08.453] Peer delivering stale transactions peer=651ba08ecbfd7d16b6a43bae82ffd54c31fb4663145c469d7c7e4b0777077374 rejected=45
WARN [12-04|12:04:10.177] Peer delivering stale transactions peer=651ba08ecbfd7d16b6a43bae82ffd54c31fb4663145c469d7c7e4b0777077374 rejected=38
WARN [12-04|12:04:12.008] Peer delivering stale transactions peer=651ba08ecbfd7d16b6a43bae82ffd54c31fb4663145c469d7c7e4b0777077374 rejected=38
WARN [12-04|12:04:12.388] Peer delivering stale transactions peer=651ba08ecbfd7d16b6a43bae82ffd54c31fb4663145c469d7c7e4b0777077374 rejected=50
INFO [12-04|12:04:12.426] Forkchoice requested sync to new head number=16,111,323 hash=88ec6d..20dd8f
WARN [12-04|12:04:12.990] Peer delivering stale transactions peer=651ba08ecbfd7d16b6a43bae82ffd54c31fb4663145c469d7c7e4b0777077374 rejected=35
WARN [12-04|12:04:14.504] Peer delivering stale transactions peer=651ba08ecbfd7d16b6a43bae82ffd54c31fb4663145c469d7c7e4b0777077374 rejected=47
WARN [12-04|12:04:15.779] Peer delivering stale transactions peer=651ba08ecbfd7d16b6a43bae82ffd54c31fb4663145c469d7c7e4b0777077374 rejected=46
INFO [12-04|12:04:24.338] Forkchoice requested sync to new head number=16,111,324 hash=5cea4e..ecd1d0
ERROR[12-04|12:04:26.120] Error in block freeze operation err="block receipts missing, can't freeze block 15606390"
INFO [12-04|12:04:27.393] Imported new chain segment number=16,111,312 hash=c6ebac..c1c3ab blocks=5 txs=948 mgas=69.596 elapsed=14.120s mgasps=4.929 age=2m28s dirty=995.57MiB ignored=1
INFO [12-04|12:04:35.997] Forkchoice requested sync to new head number=16,111,325 hash=7d3760..38cac6
INFO [12-04|12:04:47.802] Imported new chain segment number=16,111,321 hash=272ca0..25cf33 blocks=9 txs=1217 mgas=142.953 elapsed=20.409s mgasps=7.004 age=1m dirty=997.49MiB
INFO [12-04|12:04:48.731] Imported new chain segment number=16,111,323 hash=88ec6d..20dd8f blocks=2 txs=240 mgas=18.027 elapsed=928.874ms mgasps=19.408 dirty=996.39MiB
WARN [12-04|12:04:53.466] Checkpoint challenge timed out, dropping id=4f60347ed7d7af61 conn=dyndial addr=3.80.12.255:30720 type=
INFO [12-04|12:04:54.932] Imported new chain segment number=16,111,325 hash=0946fa..46f44b blocks=2 txs=253 mgas=7.600 elapsed=130.270ms mgasps=58.340 dirty=995.09MiB ignored=1
WARN [12-04|12:05:02.230] Ignoring payload with missing parent number=16,111,326 hash=3a0143..2678a4 parent=7d3760..38cac6
WARN [12-04|12:05:13.531] Ignoring payload with missing parent number=16,111,327 hash=e97873..822c73 parent=3a0143..2678a4
INFO [12-04|12:05:15.114] Forkchoice requested sync to new head number=16,111,327 hash=e97873..822c73
WARN [12-04|12:05:15.114] Beacon chain gapped head=16,111,325 newHead=16,111,327
INFO [12-04|12:05:15.256] Syncing beacon headers downloaded=90 left=0 eta=0s
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x1c0 pc=0xad6fdb]
goroutine 1555316 [running]:
github.com/ethereum/go-ethereum/eth/downloader.(*Downloader).syncWithPeer(0xc000041500, 0x0, {0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...}, ...)
github.com/ethereum/go-ethereum/eth/downloader/downloader.go:491 +0x49b
github.com/ethereum/go-ethereum/eth/downloader.(*Downloader).synchronise(0xc168424d80?, {0x0, 0x0}, {0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...}, ...)
github.com/ethereum/go-ethereum/eth/downloader/downloader.go:439 +0x512
github.com/ethereum/go-ethereum/eth/downloader.(*beaconBackfiller).resume.func1()
github.com/ethereum/go-ethereum/eth/downloader/beaconsync.go:108 +0x96
created by github.com/ethereum/go-ethereum/eth/downloader.(*beaconBackfiller).resume
github.com/ethereum/go-ethereum/eth/downloader/beaconsync.go:98 +0x136

```

--- a/eth/downloader/skeleton.go
+++ b/eth/downloader/skeleton.go
@@ -977,8 +977,14 @@ func (s *skeleton) processResponse(res *headerResponse) (linked bool, merged boo
 				// the expected new sync cycle after some propagated blocks. Log
 				// it for debugging purposes, explicitly clean and don't escalate.
 				case subchains == 2 && s.progress.Subchains[1].Head == s.progress.Subchains[1].Tail:
-					log.Debug("Cleaning previous beacon sync state", "head", s.progress.Subchains[1].Head)
-					rawdb.DeleteSkeletonHeader(batch, s.progress.Subchains[1].Head)
+					// Remove the leftover skeleton header associated with old
+					// skeleton chain only if it's not covered by the current
+					// skeleton range.
+					if s.progress.Subchains[1].Head < s.progress.Subchains[0].Tail {
+						log.Debug("Cleaning previous beacon sync state", "head", s.progress.Subchains[1].Head)
+						rawdb.DeleteSkeletonHeader(batch, s.progress.Subchains[1].Head)
+					}
+					// Drop the leftover skeleton chain since it's stale.
 					s.progress.Subchains = s.progress.Subchains[:1]
 
 				// If we have more than one header or more than one leftover chain,

--- a/eth/downloader/skeleton.go
+++ b/eth/downloader/skeleton.go
@@ -520,7 +520,7 @@ func (s *skeleton) initSync(head *types.Header) {
 				}
 				break
 			}
-			// If the last subchain can be extended, we're lucky. Otherwise create
+			// If the last subchain can be extended, we're lucky. Otherwise, create
 			// a new subchain sync task.
 			var extended bool
 			if n := len(s.progress.Subchains); n > 0 {

--- a/eth/downloader/skeleton_test.go
+++ b/eth/downloader/skeleton_test.go
@@ -37,7 +37,7 @@ import (
 type hookedBackfiller struct {
 	// suspendHook is an optional hook to be called when the filler is requested
 	// to be suspended.
-	suspendHook func()
+	suspendHook func() *types.Header
 
 	// resumeHook is an optional hook to be called when the filler is requested
 	// to be resumed.
@@ -56,7 +56,7 @@ func newHookedBackfiller() backfiller {
 // on initial startup.
 func (hf *hookedBackfiller) suspend() *types.Header {
 	if hf.suspendHook != nil {
-		hf.suspendHook()
+		return hf.suspendHook()
 	}
 	return nil // we don't really care about header cleanups for now
 }
@@ -525,9 +525,21 @@ func TestSkeletonSyncRetrievals(t *testing.T) {
 			Number:     big.NewInt(int64(i)),
 		})
 	}
+	// Some tests require a forking side chain to trigger cornercases.
+	var sidechain []*types.Header
+	for i := 0; i < len(chain)/2; i++ { // Fork at block #5000
+		sidechain = append(sidechain, chain[i])
+	}
+	for i := len(chain) / 2; i < len(chain); i++ {
+		sidechain = append(sidechain, &types.Header{
+			ParentHash: sidechain[i-1].Hash(),
+			Number:     big.NewInt(int64(i)),
+			Extra:      []byte("B"), // force a different hash
+		})
+	}
 	tests := []struct {
-		headers  []*types.Header // Database content (beside the genesis)
-		oldstate []*subchain     // Old sync state with various interrupted subchains
+		fill          bool // Whether to run a real backfiller in this test case
+		unpredictable bool // Whether to ignore drops/serves due to uncertain packet assignments
 
 		head     *types.Header       // New head header to announce to reorg to
 		peers    []*skeletonTestPeer // Initial peer set to start the sync with
@@ -760,11 +772,41 @@ func TestSkeletonSyncRetrievals(t *testing.T) {
 			endstate: []*subchain{{Head: 2*requestHeaders + 2, Tail: 1}},
 			endserve: 4 * requestHeaders,
 		},
+		// This test reproduces a bug caught by (@rjl493456442) where a skeleton
+		// header goes missing, causing the sync to get stuck and/or panic.
+		//
+		// The setup requires a previously successfully synced chain up to a block
+		// height N. That results is a single skeleton header (block N) and a single
+		// subchain (head N, Tail N) being stored on disk.
+		//
+		// The following step requires a new sync cycle to a new side chain of a
+		// height higher than N, and an ancestor lower than N (e.g. N-2, N+2).
+		// In this scenario, when processing a batch of headers, a link point of
+		// N-2 will be found, meaning that N-1 and N have been overwritten.
+		//
+		// The link event triggers an early exit, noticing that the previous sub-
+		// chain is a leftover and deletes it (with it's skeleton header N). But
+		// since skeleton header N has been overwritten to the new side chain, we
+		// end up losing it and creating a gap.
+		{
+			fill:          true,
+			unpredictable: true, // We have good and bad peer too, bad may be dropped, test too short for certainty
+
+			head:     chain[len(chain)/2+1], // Sync up until the sidechain common ancestor + 2
+			peers:    []*skeletonTestPeer{newSkeletonTestPeer("test-peer-oldchain", chain)},
+			midstate: []*subchain{{Head: uint64(len(chain)/2 + 1), Tail: 1}},
+
+			newHead:  sidechain[len(sidechain)/2+3], // Sync up until the sidechain common ancestor + 4
+			newPeer:  newSkeletonTestPeer("test-peer-newchain", sidechain),
+			endstate: []*subchain{{Head: uint64(len(sidechain)/2 + 3), Tail: uint64(len(chain) / 2)}},
+		},
 	}
 	for i, tt := range tests {
 		// Create a fresh database and initialize it with the starting state
 		db := rawdb.NewMemoryDatabase()
-		rawdb.WriteHeader(db, chain[0])
+
+		rawdb.WriteBlock(db, types.NewBlockWithHeader(chain[0]))
+		rawdb.WriteReceipts(db, chain[0].Hash(), chain[0].Number.Uint64(), types.Receipts{})
 
 		// Create a peer set to feed headers through
 		peerset := newPeerSet()
@@ -780,8 +822,43 @@ func TestSkeletonSyncRetrievals(t *testing.T) {
 			peerset.Unregister(peer)
 			dropped[peer]++
 		}
+		// Create a backfiller if we need to run more advanced tests
+		filler := newHookedBackfiller()
+		if tt.fill {
+			var filled *types.Header
+
+			filler = &hookedBackfiller{
+				resumeHook: func() {
+					var progress skeletonProgress
+					json.Unmarshal(rawdb.ReadSkeletonSyncStatus(db), &progress)
+
+					for progress.Subchains[0].Tail < progress.Subchains[0].Head {
+						header := rawdb.ReadSkeletonHeader(db, progress.Subchains[0].Tail)
+
+						rawdb.WriteBlock(db, types.NewBlockWithHeader(header))
+						rawdb.WriteReceipts(db, header.Hash(), header.Number.Uint64(), types.Receipts{})
+
+						rawdb.DeleteSkeletonHeader(db, header.Number.Uint64())
+
+						progress.Subchains[0].Tail++
+						progress.Subchains[0].Next = header.Hash()
+					}
+					filled = rawdb.ReadSkeletonHeader(db, progress.Subchains[0].Tail)
+
+					rawdb.WriteBlock(db, types.NewBlockWithHeader(filled))
+					rawdb.WriteReceipts(db, filled.Hash(), filled.Number.Uint64(), types.Receipts{})
+				},
+
+				suspendHook: func() *types.Header {
+					prev := filled
+					filled = nil
+
+					return prev
+				},
+			}
+		}
 		// Create a skeleton sync and run a cycle
-		skeleton := newSkeleton(db, peerset, drop, newHookedBackfiller())
+		skeleton := newSkeleton(db, peerset, drop, filler)
 		skeleton.Sync(tt.head, true)
 
 		var progress skeletonProgress
@@ -815,19 +892,21 @@ func TestSkeletonSyncRetrievals(t *testing.T) {
 			t.Error(err)
 			continue
 		}
-		var served uint64
-		for _, peer := range tt.peers {
-			served += atomic.LoadUint64(&peer.served)
-		}
-		if served != tt.midserve {
-			t.Errorf("test %d, mid state: served headers mismatch: have %d, want %d", i, served, tt.midserve)
-		}
-		var drops uint64
-		for _, peer := range tt.peers {
-			drops += atomic.LoadUint64(&peer.dropped)
-		}
-		if drops != tt.middrop {
-			t.Errorf("test %d, mid state: dropped peers mismatch: have %d, want %d", i, drops, tt.middrop)
+		if !tt.unpredictable {
+			var served uint64
+			for _, peer := range tt.peers {
+				served += atomic.LoadUint64(&peer.served)
+			}
+			if served != tt.midserve {
+				t.Errorf("test %d, mid state: served headers mismatch: have %d, want %d", i, served, tt.midserve)
+			}
+			var drops uint64
+			for _, peer := range tt.peers {
+				drops += atomic.LoadUint64(&peer.dropped)
+			}
+			if drops != tt.middrop {
+				t.Errorf("test %d, mid state: dropped peers mismatch: have %d, want %d", i, drops, tt.middrop)
+			}
 		}
 		// Apply the post-init events if there's any
 		if tt.newHead != nil {
@@ -868,25 +947,27 @@ func TestSkeletonSyncRetrievals(t *testing.T) {
 			continue
 		}
 		// Check that the peers served no more headers than we actually needed
-		served = 0
-		for _, peer := range tt.peers {
-			served += atomic.LoadUint64(&peer.served)
-		}
-		if tt.newPeer != nil {
-			served += atomic.LoadUint64(&tt.newPeer.served)
-		}
-		if served != tt.endserve {
-			t.Errorf("test %d, end state: served headers mismatch: have %d, want %d", i, served, tt.endserve)
-		}
-		drops = 0
-		for _, peer := range tt.peers {
-			drops += atomic.LoadUint64(&peer.dropped)
-		}
-		if tt.newPeer != nil {
-			drops += atomic.LoadUint64(&tt.newPeer.dropped)
-		}
-		if drops != tt.middrop {
-			t.Errorf("test %d, end state: dropped peers mismatch: have %d, want %d", i, drops, tt.middrop)
+		if !tt.unpredictable {
+			served := uint64(0)
+			for _, peer := range tt.peers {
+				served += atomic.LoadUint64(&peer.served)
+			}
+			if tt.newPeer != nil {
+				served += atomic.LoadUint64(&tt.newPeer.served)
+			}
+			if served != tt.endserve {
+				t.Errorf("test %d, end state: served headers mismatch: have %d, want %d", i, served, tt.endserve)
+			}
+			drops := uint64(0)
+			for _, peer := range tt.peers {
+				drops += atomic.LoadUint64(&peer.dropped)
+			}
+			if tt.newPeer != nil {
+				drops += atomic.LoadUint64(&tt.newPeer.dropped)
+			}
+			if drops != tt.enddrop {
+				t.Errorf("test %d, end state: dropped peers mismatch: have %d, want %d", i, drops, tt.middrop)
+			}
 		}
 		// Clean up any leftover skeleton sync resources
 		skeleton.Terminate()

